dbux-projects/src/backend/Backlog.js

```diff
@@ -1,73 +1,112 @@
 import { newLogger } from '@dbux/common/src/log/logger';
+import { isEqual } from 'lodash';
+import SafetyStorage from './SafetyStorage';
 
 /**
  * @typedef {import('../ProjectsManager').default} PracticeManager
  */
 
 const { log, debug, warn, error: logError } = newLogger('Backlog');
 
+const Verbose = true;
+
 const keyName = 'dbux.projects.backend.backlog';
 
-export default class Backlog {
+export default class Backlog extends SafetyStorage {
   /**
    * @param {PracticeManager} practiceManager 
+   * @param {Function} doWriteFunction
    */
   constructor(practiceManager, doWriteFunction) {
-    this.practiceManager = practiceManager;
-    this.backlog = [];
+    super(keyName);
 
-    this.init();
+    this.practiceManager = practiceManager;
     this._doWrite = doWriteFunction;
-  }
 
-  init() {
-    this.backlog = this.practiceManager.externals.storage.get(keyName) || [];
-    debug(`Backlog init: `, this.backlog);
+    Verbose && debug('Backlog init: ', this.get());
   }
 
-  async save() {
-    return this.practiceManager.externals.storage.set(keyName, this.backlog);
+  safeGet() {
+    return this.get() || [];
   }
 
   size() {
-    return !!this.backlog.length;
+    return !!this.safeGet().length;
   }
 
   /**
    * Remember write action and try again later.
    * @param {object} writeRequest
    */
   async add(writeRequest) {
-    this.backlog.push(writeRequest);
-    await this.save();
+    await this.acquireLock();
+
+    let backlog = this.safeGet();
+    backlog.push(writeRequest);
+    await this.set(backlog);
+
+    this.releaseLock();
+
+    Verbose && debug('contents after add', writeRequest, this.get());
   }
 
   /**
    * If backlog is corrupted, allow user to reset everything.
    */
   async resetBacklog() {
-    this.backlog = [];
-    await this.save();
+    await this.acquireLock();
+    await this.set([]);
+    this.releaseLock();
+
+    Verbose && debug('reset all contents');
   }
 
   // async _doWrite(writeRequest) {
   //   
   // }
 
   async replay() {
-    while (this.size()) {
-      let writeRequest = this.backlog[0];
+    await this.acquireLock();
+    let backlog = this.safeGet();
+    Verbose && debug('replay');
+
+    while (backlog.length) {
+      let writeRequest = backlog[0];
+      Verbose && debug('replay request', writeRequest);
 
       try {
         await this._doWrite(writeRequest);
-        this.backlog.shift();
-        await this.save();
+        backlog.shift();
       }
       catch (err) {
-        this.backlog.shift();
-        await this.save();
+        backlog.shift();
         warn(`Removed write request after error: ${err.message}\n\n${JSON.stringify(writeRequest)}`);
       }
```

Comment:
The storage will be locked before the loop, no need to worry about new data been pushed while writing, thus we can make this loop O(1) (use for-of or forEach) instead.